File: apps/api-rs/src/document/markdown.rs
use super::plain_text::find_text_char_range;
use super::shared::{attach_table_cell_anchors, finalize_parsed, normalize_space, table_markdown};
use super::text_utils::decode_text;
use super::*;
pub(super) fn parse_markdown(
    doc_id: Uuid,
    parse_job_id: Uuid,
    title: &str,
    bytes: &[u8],
) -> Result<ParsedDocument> {
    let text = decode_text(bytes).context("invalid_markdown_encoding")?;
    let mut blocks = Vec::new();
    let mut tables = Vec::new();
    let mut heading_path: Vec<(i32, String)> = Vec::new();
    let mut in_frontmatter = false;
    let mut frontmatter = Vec::new();
    let mut in_code = false;
    let mut code_fence = Vec::new();
    let mut paragraph = Vec::new();
    let mut table_lines = Vec::new();
    let mut warnings = Vec::new();

    for (line_idx, line) in text.lines().enumerate() {
        let trimmed = line.trim();
        if line_idx == 0 && trimmed == "---" {
            in_frontmatter = true;
            continue;
        }
        if in_frontmatter {
            if trimmed == "---" {
                in_frontmatter = false;
                if !frontmatter.is_empty() {
                    push_markdown_block(
                        &mut blocks,
                        "metadata",
                        frontmatter.join("\n"),
                        None,
                        &heading_path,
                        json!({"format": "md", "source": "frontmatter"}),
                    );
                    frontmatter.clear();
                }
            } else {
                frontmatter.push(line.to_string());
            }
            continue;
        }

        if trimmed.starts_with("```") || trimmed.starts_with("~~~") {
            if in_code {
                code_fence.push(line.to_string());
                push_markdown_block(
                    &mut blocks,
                    "code",
                    code_fence.join("\n"),
                    None,
                    &heading_path,
                    json!({"format": "md", "source": "code_fence"}),
                );
                code_fence.clear();
                in_code = false;
            } else {
                flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
                flush_markdown_table(
                    doc_id,
                    parse_job_id,
                    &mut blocks,
                    &mut tables,
                    &mut table_lines,
                    &heading_path,
                );
                in_code = true;
                code_fence.push(line.to_string());
            }
            continue;
        }
        if in_code {
            code_fence.push(line.to_string());
            continue;
        }

        if let Some((level, heading)) = markdown_heading(trimmed) {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
            flush_markdown_table(
                doc_id,
                parse_job_id,
                &mut blocks,
                &mut tables,
                &mut table_lines,
                &heading_path,
            );
            heading_path.retain(|(existing, _)| *existing < level);
            heading_path.push((level, heading.clone()));
            push_markdown_block(
                &mut blocks,
                "heading",
                heading,
                Some(level),
                &heading_path[..heading_path.len().saturating_sub(1)],
                json!({"format": "md", "line": line_idx + 1}),
            );
            continue;
        }

        if trimmed.starts_with("<!--") {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
            push_markdown_block(
                &mut blocks,
                "comment",
                trimmed.to_string(),
                None,
                &heading_path,
                json!({"format": "md", "line": line_idx + 1}),
            );
            continue;
        }

        if markdown_table_line(trimmed) {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
            table_lines.push(trimmed.to_string());
            continue;
        } else {
            flush_markdown_table(
                doc_id,
                parse_job_id,
                &mut blocks,
                &mut tables,
                &mut table_lines,
                &heading_path,
            );
        }

        if trimmed.is_empty() {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
        } else if markdown_list_item(trimmed) {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
            push_markdown_block(
                &mut blocks,
                "list_item",
                trimmed.to_string(),
                None,
                &heading_path,
                json!({"format": "md", "line": line_idx + 1}),
            );
        } else if trimmed.starts_with('>') {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
            push_markdown_block(
                &mut blocks,
                "blockquote",
                trimmed.trim_start_matches('>').trim().to_string(),
                None,
                &heading_path,
                json!({"format": "md", "line": line_idx + 1}),
            );
        } else if trimmed.starts_with("![") {
            flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
            push_markdown_block(
                &mut blocks,
                "image",
                trimmed.to_string(),
                None,
                &heading_path,
                json!({"format": "md", "line": line_idx + 1}),
            );
        } else {
            paragraph.push(trimmed.to_string());
        }
    }

    if in_frontmatter {
        push_markdown_block(
            &mut blocks,
            "metadata",
            frontmatter.join("\n"),
            None,
            &heading_path,
            json!({"format": "md", "source": "frontmatter"}),
        );
        warnings.push("markdown_unclosed_frontmatter".to_string());
    }
    if in_code {
        push_markdown_block(
            &mut blocks,
            "code",
            code_fence.join("\n"),
            None,
            &heading_path,
            json!({"format": "md", "source": "code_fence"}),
        );
        warnings.push("markdown_unclosed_code_fence".to_string());
    }
    flush_markdown_paragraph(&mut blocks, &mut paragraph, &heading_path);
    flush_markdown_table(
        doc_id,
        parse_job_id,
        &mut blocks,
        &mut tables,
        &mut table_lines,
        &heading_path,
    );

    let mut anchors = Vec::new();
    attach_table_cell_anchors(
        doc_id,
        parse_job_id,
        "md",
        &mut blocks,
        &tables,
        &mut anchors,
    );
    let mut char_cursor = 0usize;
    for block in &mut blocks {
        if block.anchor_ids.is_empty() {
            let mut anchor = SourceAnchor::structural(
                doc_id,
                parse_job_id,
                Uuid::nil(),
                "md",
                &block.block_type,
                block.block_id,
                block.page_start,
                block.slide_index,
                block.source_ref.clone(),
                &block.text,
            );
            if let Some((start, end)) = find_text_char_range(&text, &block.text, &mut char_cursor) {
                anchor.char_range = Some(CharRange { start, end });
            }
            block.anchor_ids.push(anchor.anchor_id);
            anchors.push(anchor);
        }
    }

    let mut parsed = finalize_parsed(
        doc_id,
        parse_job_id,
        "md",
        title,
        None,
        blocks,
        tables,
        anchors,
    );
    parsed.warnings.extend(warnings);
    Ok(parsed)
}

fn markdown_heading(line: &str) -> Option<(i32, String)> {
    let hashes = line.chars().take_while(|c| *c == '#').count();
    if hashes == 0 || hashes > 6 {
        return None;
    }
    let rest = line.get(hashes..)?.trim();
    if rest.is_empty() {
        None
    } else {
        Some((hashes as i32, rest.trim_matches('#').trim().to_string()))
    }
}

fn markdown_list_item(line: &str) -> bool {
    line.starts_with("- ")
        || line.starts_with("* ")
        || line.starts_with("+ ")
        || line
            .split_once(". ")
            .map(|(prefix, _)| !prefix.is_empty() && prefix.chars().all(|c| c.is_ascii_digit()))
            .unwrap_or(false)
}

fn markdown_table_line(line: &str) -> bool {
    line.starts_with('|') && line.ends_with('|') && line.matches('|').count() >= 2
}

fn push_markdown_block(
    blocks: &mut Vec<ParsedBlock>,
    block_type: &str,
    text: String,
    heading_level: Option<i32>,
    heading_path: &[(i32, String)],
    metadata: Value,
) {
    if text.trim().is_empty() {
        return;
    }
    blocks.push(ParsedBlock {
        block_id: Uuid::new_v4(),
        block_index: blocks.len() as i32,
        block_type: block_type.to_string(),
        text,
        heading_level,
        heading_path: heading_path.iter().map(|(_, h)| h.clone()).collect(),
        page_start: None,
        page_end: None,
        slide_index: None,
        table_id: None,
        bbox: None,
        anchor_ids: vec![],
        source_ref: json!({"format": "md", "index": blocks.len()}),
        metadata,
    });
}

fn flush_markdown_paragraph(
    blocks: &mut Vec<ParsedBlock>,
    paragraph: &mut Vec<String>,
    heading_path: &[(i32, String)],
) {
    if paragraph.is_empty() {
        return;
    }
    let text = paragraph.join("\n");
    paragraph.clear();
    push_markdown_block(
        blocks,
        "paragraph",
        text,
        None,
        heading_path,
        json!({"format": "md", "source": "paragraph"}),
    );
}

fn flush_markdown_table(
    _doc_id: Uuid,
    _parse_job_id: Uuid,
    blocks: &mut Vec<ParsedBlock>,
    tables: &mut Vec<ParsedTable>,
    table_lines: &mut Vec<String>,
    heading_path: &[(i32, String)],
) {
    if table_lines.len() < 2 || !markdown_table_separator(&table_lines[1]) {
        if !table_lines.is_empty() {
            let text = table_lines.join("\n");
            push_markdown_block(
                blocks,
                "paragraph",
                text,
                None,
                heading_path,
                json!({"format": "md", "source": "paragraph"}),
            );
        }
        table_lines.clear();
        return;
    }

    let rows = table_lines
        .iter()
        .filter(|line| !line.chars().all(|c| matches!(c, '|' | '-' | ':' | ' ')))
        .map(|line| {
            line.trim_matches('|')
                .split('|')
                .map(|cell| cell.trim().to_string())
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();
    table_lines.clear();
    if rows.is_empty() {
        return;
    }

    let block_id = Uuid::new_v4();
    let table_id = Uuid::new_v4();
    let headers = rows.first().cloned().unwrap_or_default();
    let body = rows.get(1..).unwrap_or(&[]).to_vec();
    let markdown = table_markdown(&headers, &body);
    let mut cells = Vec::new();
    for (row_idx, row) in rows.iter().enumerate() {
        for (col_idx, text) in row.iter().enumerate() {
            cells.push(ParsedTableCell {
                cell_id: Uuid::new_v4(),
                row_index: row_idx as i32,
                col_index: col_idx as i32,
                rowspan: 1,
                colspan: 1,
                text: text.clone(),
                normalized_text: normalize_space(text),
                is_header: row_idx == 0,
                data_type: "text".to_string(),
                bbox: None,
                style: json!({}),
                source_ref: json!({"format": "md", "table_index": tables.len(), "row": row_idx, "col": col_idx}),
            });
        }
    }
    let path = heading_path
        .iter()
        .map(|(_, h)| h.clone())
        .collect::<Vec<_>>();
    tables.push(ParsedTable {
        table_id,
        block_id,
        table_index: tables.len() as i32,
        title: path.last().cloned(),
        heading_path: path.clone(),
        page_start: None,
        page_end: None,
        slide_index: None,
        headers: headers.clone(),
        rows: body,
        cells,
        markdown: markdown.clone(),
        quality: json!({"header_confidence": 0.9, "grid_confidence": 0.9, "warnings": []}),
        source_ref: json!({"format": "md", "table_index": tables.len()}),
    });
    blocks.push(ParsedBlock {
        block_id,
        block_index: blocks.len() as i32,
        block_type: "table".to_string(),
        text: markdown,
        heading_level: None,
        heading_path: path,
        page_start: None,
        page_end: None,
        slide_index: None,
        table_id: Some(table_id),
        bbox: None,
        anchor_ids: vec![],
        source_ref: json!({"format": "md", "node": "table", "index": tables.len()}),
        metadata: json!({"format": "md", "source": "table"}),
    });
}

fn markdown_table_separator(line: &str) -> bool {
    let cells = line.trim().trim_matches('|').split('|');
    let mut count = 0usize;
    for cell in cells {
        let cell = cell.trim().trim_matches(':');
        if cell.len() < 3 || !cell.chars().all(|ch| ch == '-') {
            return false;
        }
        count += 1;
    }
    count > 0
}

File: apps/api-bun/src/document/markdown.ts
// 移植自 apps/api-rs/src/document/markdown.rs —— 行为对齐 Rust 原版，错误信息保持一致

// Markdown 解析：frontmatter / 代码围栏 / 标题 / 列表 / 引用 / 图片 / 表格 / 段落

import type { SourceAnchor } from '../models/source_anchor.ts';
import { sourceAnchorStructural } from '../models/source_anchor.ts';
import { newUuid } from '../infra/uuid.ts';
import { findTextCharRange, type CharCursor } from './plain_text.ts';
import {
  attachTableCellAnchors,
  finalizeParsed,
  normalizeSpace,
  tableMarkdown,
} from './shared.ts';
import { decodeText, rustLines, trimRust } from './text_utils.ts';
import type { ParsedBlock, ParsedDocument, ParsedTable, ParsedTableCell } from './types.ts';
import { NIL_UUID } from './types.ts';

export function parseMarkdown(
  docId: string,
  parseJobId: string,
  title: string,
  bytes: Uint8Array,
): ParsedDocument {
  let text: string;
  try {
    text = decodeText(bytes);
  } catch (error) {
    if (error instanceof Error) {
      throw new Error('invalid_markdown_encoding:' + error.message);
    }
    throw error;
  }
  const blocks: ParsedBlock[] = [];
  const tables: ParsedTable[] = [];
  const headingPath: Array<[number, string]> = [];
  let inFrontmatter = false;
  let frontmatter: string[] = [];
  let inCode = false;
  const codeFence: string[] = [];
  const paragraph: string[] = [];
  const tableLines: string[] = [];
  const warnings: string[] = [];

  for (const [lineIdx, line] of rustLines(text).entries()) {
    const trimmed = trimRust(line);
    if (lineIdx === 0 && trimmed === '---') {
      inFrontmatter = true;
      continue;
    }
    if (inFrontmatter) {
      if (trimmed === '---') {
        inFrontmatter = false;
        if (frontmatter.length > 0) {
          pushMarkdownBlock(
            blocks,
            'metadata',
            frontmatter.join('\n'),
            null,
            headingPath,
            { format: 'md', source: 'frontmatter' },
          );
          frontmatter = [];
        }
      } else {
        frontmatter.push(line);
      }
      continue;
    }

    if (trimmed.startsWith('```') || trimmed.startsWith('~~~')) {
      if (inCode) {
        codeFence.push(line);
        pushMarkdownBlock(
          blocks,
          'code',
          codeFence.join('\n'),
          null,
          headingPath,
          { format: 'md', source: 'code_fence' },
        );
        codeFence.length = 0;
        inCode = false;
      } else {
        flushMarkdownParagraph(blocks, paragraph, headingPath);
        flushMarkdownTable(blocks, tables, tableLines, headingPath);
        inCode = true;
        codeFence.push(line);
      }
      continue;
    }
    if (inCode) {
      codeFence.push(line);
      continue;
    }

    const heading = markdownHeading(trimmed);
    if (heading !== null) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
      flushMarkdownTable(blocks, tables, tableLines, headingPath);
      const [level, headingText] = heading;
      for (let i = headingPath.length - 1; i >= 0; i -= 1) {
        if (headingPath[i]![0] >= level) headingPath.splice(i, 1);
      }
      headingPath.push([level, headingText]);
      pushMarkdownBlock(
        blocks,
        'heading',
        headingText,
        level,
        headingPath.slice(0, Math.max(headingPath.length - 1, 0)),
        { format: 'md', line: lineIdx + 1 },
      );
      continue;
    }

    if (trimmed.startsWith('<!--')) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
      pushMarkdownBlock(
        blocks,
        'comment',
        trimmed,
        null,
        headingPath,
        { format: 'md', line: lineIdx + 1 },
      );
      continue;
    }

    if (markdownTableLine(trimmed)) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
      tableLines.push(trimmed);
      continue;
    } else {
      flushMarkdownTable(blocks, tables, tableLines, headingPath);
    }

    if (trimmed.length === 0) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
    } else if (markdownListItem(trimmed)) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
      pushMarkdownBlock(
        blocks,
        'list_item',
        trimmed,
        null,
        headingPath,
        { format: 'md', line: lineIdx + 1 },
      );
    } else if (trimmed.startsWith('>')) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
      pushMarkdownBlock(
        blocks,
        'blockquote',
        trimRust(trimmed.replace(/^>+/, '')),
        null,
        headingPath,
        { format: 'md', line: lineIdx + 1 },
      );
    } else if (trimmed.startsWith('![')) {
      flushMarkdownParagraph(blocks, paragraph, headingPath);
      pushMarkdownBlock(
        blocks,
        'image',
        trimmed,
        null,
        headingPath,
        { format: 'md', line: lineIdx + 1 },
      );
    } else {
      paragraph.push(trimmed);
    }
  }

  if (inFrontmatter) {
    pushMarkdownBlock(
      blocks,
      'metadata',
      frontmatter.join('\n'),
      null,
      headingPath,
      { format: 'md', source: 'frontmatter' },
    );
    warnings.push('markdown_unclosed_frontmatter');
  }
  if (inCode) {
    pushMarkdownBlock(
      blocks,
      'code',
      codeFence.join('\n'),
      null,
      headingPath,
      { format: 'md', source: 'code_fence' },
    );
    warnings.push('markdown_unclosed_code_fence');
  }
  flushMarkdownParagraph(blocks, paragraph, headingPath);
  flushMarkdownTable(blocks, tables, tableLines, headingPath);

  const anchors: SourceAnchor[] = [];
  attachTableCellAnchors(docId, parseJobId, 'md', blocks, tables, anchors);
  const charCursor: CharCursor = { pos: 0 };
  for (const block of blocks) {
    if (block.anchor_ids.length > 0) continue;
    const anchor = sourceAnchorStructural(
      docId,
      parseJobId,
      NIL_UUID,
      'md',
      block.block_type,
      block.block_id,
      block.page_start,
      block.slide_index,
      block.source_ref,
      block.text,
    );
    const range = findTextCharRange(text, block.text, charCursor);
    if (range !== null) {
      anchor.char_range = { start: range[0], end: range[1] };
    }
    block.anchor_ids.push(anchor.anchor_id);
    anchors.push(anchor);
  }

  const parsed = finalizeParsed(docId, parseJobId, 'md', title, null, blocks, tables, anchors);
  parsed.warnings.push(...warnings);
  return parsed;
}

function markdownHeading(line: string): [number, string] | null {
  let hashes = 0;
  for (const ch of line) {
    if (ch !== '#') break;
    hashes += 1;
  }
  if (hashes === 0 || hashes > 6) return null;
  const rest = trimRust(line.slice(hashes));
  if (rest.length === 0) return null;
  return [hashes, trimChars(rest, '#').trim()];
}

function markdownListItem(line: string): boolean {
  if (line.startsWith('- ') || line.startsWith('* ') || line.startsWith('+ ')) return true;
  const dot = line.indexOf('. ');
  if (dot < 0) return false;
  const prefix = line.slice(0, dot);
  if (prefix.length === 0) return false;
  return [...prefix].every((ch) => ch >= '0' && ch <= '9');
}

function markdownTableLine(line: string): boolean {
  return line.startsWith('|') && line.endsWith('|') && countChar(line, '|') >= 2;
}

function pushMarkdownBlock(
  blocks: ParsedBlock[],
  blockType: string,
  blockText: string,
  headingLevel: number | null,
  headingPath: Array<[number, string]>,
  metadata: unknown,
): void {
  if (trimRust(blockText).length === 0) return;
  blocks.push({
    block_id: newUuid(),
    block_index: blocks.length,
    block_type: blockType,
    text: blockText,
    heading_level: headingLevel,
    heading_path: headingPath.map(([, heading]) => heading),
    page_start: null,
    page_end: null,
    slide_index: null,
    table_id: null,
    bbox: null,
    anchor_ids: [],
    source_ref: { format: 'md', index: blocks.length },
    metadata,
  });
}

function flushMarkdownParagraph(
  blocks: ParsedBlock[],
  paragraph: string[],
  headingPath: Array<[number, string]>,
): void {
  if (paragraph.length === 0) return;
  const text = paragraph.join('\n');
  paragraph.length = 0;
  pushMarkdownBlock(
    blocks,
    'paragraph',
    text,
    null,
    headingPath,
    { format: 'md', source: 'paragraph' },
  );
}

function flushMarkdownTable(
  blocks: ParsedBlock[],
  tables: ParsedTable[],
  tableLines: string[],
  headingPath: Array<[number, string]>,
): void {
  if (tableLines.length < 2 || !markdownTableSeparator(tableLines[1] ?? '')) {
    if (tableLines.length > 0) {
      const text = tableLines.join('\n');
      pushMarkdownBlock(
        blocks,
        'paragraph',
        text,
        null,
        headingPath,
        { format: 'md', source: 'paragraph' },
      );
    }
    tableLines.length = 0;
    return;
  }

  const rows = tableLines
    .filter((line) => ![...line].every((ch) => ch === '|' || ch === '-' || ch === ':' || ch === ' '))
    .map((line) =>
      trimChars(line, '|').split('|').map((cell) => trimRust(cell)),
    );
  tableLines.length = 0;
  if (rows.length === 0) return;

  const blockId = newUuid();
  const tableId = newUuid();
  const headers = rows[0] ?? [];
  const body = rows.slice(1);
  const markdown = tableMarkdown(headers, body);
  const cells: ParsedTableCell[] = [];
  for (const [rowIdx, row] of rows.entries()) {
    for (const [colIdx, cellText] of row.entries()) {
      cells.push({
        cell_id: newUuid(),
        row_index: rowIdx,
        col_index: colIdx,
        rowspan: 1,
        colspan: 1,
        text: cellText,
        normalized_text: normalizeSpace(cellText),
        is_header: rowIdx === 0,
        data_type: 'text',
        bbox: null,
        style: {},
        source_ref: { format: 'md', table_index: tables.length, row: rowIdx, col: colIdx },
      });
    }
  }
  const path = headingPath.map(([, heading]) => heading);
  const tableIndex = tables.length;
  tables.push({
    table_id: tableId,
    block_id: blockId,
    table_index: tableIndex,
    title: path.length > 0 ? path[path.length - 1]! : null,
    heading_path: [...path],
    page_start: null,
    page_end: null,
    slide_index: null,
    headers: [...headers],
    rows: body,
    cells,
    markdown,
    quality: { header_confidence: 0.9, grid_confidence: 0.9, warnings: [] },
    source_ref: { format: 'md', table_index: tableIndex },
  });
  // 对齐 Rust：tables.push 之后 blocks.push，source_ref.index 为新 tables.len()
  blocks.push({
    block_id: blockId,
    block_index: blocks.length,
    block_type: 'table',
    text: markdown,
    heading_level: null,
    heading_path: path,
    page_start: null,
    page_end: null,
    slide_index: null,
    table_id: tableId,
    bbox: null,
    anchor_ids: [],
    source_ref: { format: 'md', node: 'table', index: tables.length },
    metadata: { format: 'md', source: 'table' },
  });
}

function markdownTableSeparator(line: string): boolean {
  let count = 0;
  for (const rawCell of trimChars(trimRust(line), '|').split('|')) {
    const cell = trimChars(trimRust(rawCell), ':');
    if (cell.length < 3 || ![...cell].every((ch) => ch === '-')) return false;
    count += 1;
  }
  return count > 0;
}

function trimChars(value: string, ch: string): string {
  let start = 0;
  let end = value.length;
  while (start < end && value[start] === ch) start += 1;
  while (end > start && value[end - 1] === ch) end -= 1;
  return value.slice(start, end);
}

function countChar(value: string, ch: string): number {
  let count = 0;
  for (const c of value) {
    if (c === ch) count += 1;
  }
  return count;
}
